transport/p2p/src/swarm.rs

```diff
@@ -505,10 +505,15 @@ impl HoprSwarmWithProcessors {
                         match event {
                             crate::behavior::discovery::Event::NewPeerMultiddress(peer, multiaddress) => {
                                 info!(%peer, address = %multiaddress, "New record");
+                                // We add the multiaddr before the first dial
                                 swarm.add_peer_address(peer, multiaddress.clone());
 
                                 if let Err(e) = swarm.dial(peer) {
                                     error!(%peer, address = %multiaddress, error = %e, "Failed to dial the peer");
+
+                                    // If the dial fails we need to re-add the multiaddr to ensure
+                                    // the next dial attempt could work.
+                                    swarm.add_peer_address(peer, multiaddress.clone());
                                 }
```

Comment:
This could be removed altogether. The dial is not necessary anymore, since on each turn the heartbeat dials and libp2p_request_response dials automatically, if the peer is not dialed.